### PR TITLE
Make llama attention stateless

### PR DESCRIPTION
**Pull request by yangw1234**
## Description

Make llama and chatglm attention stateless (similar to chatglm2)

### CODE REVIEW

**Comment by leonardozcm** [comment]
It is cool to allocate a big kvcache as a memory pool and assign items one by one, but I doubt the way you view kv cache storage `as_strided` may result in non-contiguous memory access, increasing the addressing overhead？
I made some experiments to simulate the behavior of matrix multiplication in the Llama2 13B attention:
```python
import torch
import time
from tqdm import tqdm

# pow_n=10
# shape = [2^pow_n] * 4

bs = 1
iteration = 100
hidden_size=5120

def test_contiguous():
    T_s=[bs,64,288,hidden_size]
    W_s=[hidden_size, hidden_size]

    count = 0
    for _ in tqdm(range(iteration)):
        a = torch.empty(T_s, dtype=torch.float)
        b = torch.empty(W_s, dtype=torch.float)
        st = time.perf_counter()
        _ = torch.matmul(a, b)
        during = time.perf_counter() - st
        count+=during

    print(f"test contiguous: {count}")
    return count

def test_blocked():
    T_s=[bs, 64, 576, hidden_size]
    W_s=[hidden_size, hidden_size]

    count = 0
    for _ in tqdm(range(iteration)):
        a_0:torch.Tensor = torch.empty(T_s, dtype=torch.float)
        a = a_0.as_strided([1, 64, 288, hidden_size], stride=a_0.stride())
        b = torch.empty(W_s, dtype=torch.float)
        st = time.perf_counter()
        _ = torch.matmul(a, b)
        during = time.perf_counter() - st
        count+=during

    print(f"test blocked: {count}")
    return count

blocked_timing = 0
contiguous_timing = 0

for _ in range(10):

    blocked_timing+= test_blocked()
    contiguous_timing+= test_contiguous()

print(f"-{(blocked_timing-contiguous_timing)*100.0/contiguous_timing:0.2f}%")

```
The results indicate a potential performance loss of 7.56%:
```log
(changmim-llm) cpx@cpx-2:/disk1/changmin/BigDL$ numactl -C 0-27 -m 0 python test_multi.py
100%|████████████████████████████████████████████████████████████████████| 100/100 [00:38<00:00,  2.62it/s]
test blocked: 38.15154311899096
100%|████████████████████████████████████████████████████████████████████| 100/100 [00:35<00:00,  2.81it/s]

...

-7.56%
```

**Comment by jason-dai** [comment]
> It is cool to allocate a big kvcache as a memory pool and assign items one by one, but I doubt the way you view kv cache storage `as_strided` may result in non-contiguous memory access, increasing the addressing overhead？ I made some experiments to simulate the behavior of matrix multiplication in the Llama2 13B attention:
> 
> ```python
> import torch
> import time
> from tqdm import tqdm
> 
> # pow_n=10
> # shape = [2^pow_n] * 4
> 
> bs = 1
> iteration = 100
> hidden_size=5120
> 
> def test_contiguous():
>     T_s=[bs,64,288,hidden_size]
>     W_s=[hidden_size, hidden_size]
> 
>     count = 0
>     for _ in tqdm(range(iteration)):
>         a = torch.empty(T_s, dtype=torch.float)
>         b = torch.empty(W_s, dtype=torch.float)
>         st = time.perf_counter()
>         _ = torch.matmul(a, b)
>         during = time.perf_counter() - st
>         count+=during
> 
>     print(f"test contiguous: {count}")
>     return count
> 
> def test_blocked():
>     T_s=[bs, 64, 576, hidden_size]
>     W_s=[hidden_size, hidden_size]
> 
>     count = 0
>     for _ in tqdm(range(iteration)):
>         a_0:torch.Tensor = torch.empty(T_s, dtype=torch.float)
>         a = a_0.as_strided([1, 64, 288, hidden_size], stride=a_0.stride())
>         b = torch.empty(W_s, dtype=torch.float)
>         st = time.perf_counter()
>         _ = torch.matmul(a, b)
>         during = time.perf_counter() - st
>         count+=during
> 
>     print(f"test blocked: {count}")
>     return count
> 
> blocked_timing = 0
> contiguous_timing = 0
> 
> for _ in range(10):
> 
>     blocked_timing+= test_blocked()
>     contiguous_timing+= test_contiguous()
> 
> print(f"-{(blocked_timing-contiguous_timing)*100.0/contiguous_timing:0.2f}%")
> ```
> 
> The results indicate a potential performance loss of 7.56%:
> 
> ```
> (changmim-llm) cpx@cpx-2:/disk1/changmin/BigDL$ numactl -C 0-27 -m 0 python test_multi.py
> 100%|████████████████████████████████████████████████████████████████████| 100/100 [00:38<00:00,  2.62it/s]
> test blocked: 38.15154311899096
> 100%|████████████████████████████████████████████████████████████████████| 100/100 [00:35<00:00,  2.81it/s]
> test contiguous: 35.49933813116513
> 100%|████████████████████████████████████████████████████████████████████| 100/100 [00:37<00:00,  2.67it/s]
> test blocked: 37.478191399015486
> 100%|████████████████████████████████████████████████████████████████████| 100/100 [00:35<00:00,  2.82it/s]
> test contiguous: 35.428268634248525
> ...
> 
> -7.56%
> ```

Any idea to fix it?

**Comment by jason-dai** [comment]
> This is an efficiency issue and does not impact the final results. huggingface llama_attention just concat the old ones with fresh kv states:
> 
> ```python
>         if past_key_value is not None:
>             # reuse k, v, self_attention
>             key_states = torch.cat([past_key_value[0], key_states], dim=2)
>             value_states = torch.cat([past_key_value[1], value_states], dim=2)
> ```
> 
> We can allocate a contiguous new tensor upon the old one during every call, but this approach still remain overhead. Actually I just observe this a potential issue, and it is hard to make increasing kv cache contiguous. Does this really matter @yangw1234 ?

concating or allocating new tensor is too slow

**Comment by yangw1234** [comment]
@leonardozcm 
the approach in this PR should be the same as before right? 

https://github.com/intel-analytics/BigDL/pull/8928/files#diff-d929bb7b0da67cf943d9d4b419b2dfe395d9a407cc5c406285c056bf5ff5dad4L80-L81 does not result in continuous  tensors either. 

You can set optimize_model=False get the default behavior of huggingface implementation. Removing the memory copy in `key_states = torch.cat([past_key_value[0], key_states], dim=2)` is reason why we have this file.

**Comment by yangw1234** [comment]
> We can allocate a contiguous new tensor upon the old one during every call, but this approach still remain overhead. Actually I just observe this a potential issue, and it is hard to make increasing kv cache contiguous. Does this really matter @yangw1234 ?

"Increasing kv cache contiguous" is possible, but we would need customized attention kernels for that. See [paged attention](https://vllm.ai/) and[ tokenattention](https://github.com/ModelTC/lightllm/blob/main/docs/TokenAttention.md)

**Review by leonardozcm — APPROVED** [review]
